### PR TITLE
Feature/masq 439 update the http connector for pot object 

### PR DESCRIPTION
**Pull request by gastonNutmeg**

### CODE REVIEW

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java [hunk 1]
```diff
@@ -24,14 +25,7 @@
...
 import java.time.Clock;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Base64;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (gkalogiros-nutmeg) [inline src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: not sure if you have any coding standards for this but having explicit imports is a better pattern to follow

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java [hunk 2]
```diff
@@ -98,14 +93,13 @@ public void write(final Collection<SinkRecord> records)
...
 
             // build batch key
-            String formattedKeyPattern = httpSinkConfig.batchKeyPattern
-                  .replace("${key}", record.key() == null ? "" : record.key().toString())
-                  .replace("${topic}", record.topic());
+            String formattedKeyPattern = evaluateReplacements(httpSinkConfig.batchKeyPattern,
+                  record);
 
             // add to batch and check for batch size limit
             if (!batches.containsKey(formattedKeyPattern)) {
```
Comment (gkalogiros-nutmeg) [inline src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: You can replace this if/else statement with 
```
batches.getOrDefault(formattedKeyPattern, new ArrayList<>()).add(record)
```

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java [hunk 3]
```diff
@@ -166,44 +156,54 @@ private void sendBatch(String formattedKeyPattern)
...
 
-    private static String buildJsonFromStruct(Struct struct) {
+    private String evaluateReplacements(String inputString, SinkRecord record) {
+        return inputString
+              .replace("${key}", getKey(record))
+              .replace("${topic}", record.topic());
+    }
+
+    private String getKey(SinkRecord record) {
+        return record.key() == null ? StringUtils.EMPTY : StringUtils.trim(record.key().toString());
```
Comment (gkalogiros-nutmeg) [inline src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: You can do it the java8 way -> Optional.ofNullable(record.key()).map(k -> StringUtils.trim(record.key().toString())).orElse(StringUtils.EMPTY)

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java [hunk 4]
```diff
@@ -114,10 +108,9 @@ public void write(final Collection<SinkRecord> records)
             }
         }
         flushBatches();
-
     }
 
-    public void flushBatches()
+    private void flushBatches()
```
Comment (gkalogiros-nutmeg) [inline src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: Not part of this change but all private methods need to be below the public ones. I'd consider moving `extractPrivateKeyFromConfig` down here

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java [hunk 5]
```diff
@@ -129,21 +122,18 @@ private void sendBatch(String formattedKeyPattern)
...
+        SinkRecord record = records.get(0);
+        String formattedUrl = evaluateReplacements(httpSinkConfig.httpApiUrl, record);
         HttpSinkConfig.RequestMethod requestMethod = httpSinkConfig.requestMethod;
 
         // add headers
         Map<String, String> headers = Arrays
               .stream(httpSinkConfig.headers.split(httpSinkConfig.headerSeparator))
-              .filter((s) -> s.contains(":"))
-              .map((s) -> s.split(":"))
+              .filter((s) -> s.contains(HEADER_VALUE_SEPARATOR))
```
Comment (gkalogiros-nutmeg) [inline src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: in single parameter functions you don't need to wrap the input within parenthesis `s -> s.contains(HEADER....)`

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java [hunk 6]
```diff
@@ -166,44 +156,54 @@ private void sendBatch(String formattedKeyPattern)
...
             HttpResponse httpResponse = new HttpResponse(
                   response.getStatusCode(),
                   formattedUrl,
-                  response.getStatusMessage() == null ? StringUtils.EMPTY
+                  response.getStatusMessage() == null
+                        ? StringUtils.EMPTY
                         : response.getStatusMessage(),
                   response.getBody() == null ? StringUtils.EMPTY : response.getBody());
-            kafkaClient.publish(recordKey, httpSinkConfig.responseTopic, httpResponse);
+            kafkaClient.publish(getKey(record), httpSinkConfig.responseTopic, httpResponse);
```
Comment (gkalogiros-nutmeg) [inline src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: I'm wondering if you need to publish the "empty key" records here. If you dive into the `getKey` method, it may return an empty key right? Do we want to publish these entries?

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java [hunk 7]
```diff
@@ -53,6 +47,7 @@
     private final HttpSinkConfig httpSinkConfig;
     private static final Logger log = LoggerFactory.getLogger(HttpApiWriter.class);
     private Map<String, List<SinkRecord>> batches = new HashMap<>();
+    private static final String HEADER_VALUE_SEPARATOR = ":";
```
Comment (gkalogiros-nutmeg) [inline src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: That's subjective but in most code bases, global variables are at the top of the class.

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java [hunk 8]
```diff
@@ -166,44 +156,54 @@ private void sendBatch(String formattedKeyPattern)
...
+        return inputString
+              .replace("${key}", getKey(record))
+              .replace("${topic}", record.topic());
+    }
+
+    private String getKey(SinkRecord record) {
+        return record.key() == null ? StringUtils.EMPTY : StringUtils.trim(record.key().toString());
+    }
+
+    private String buildJsonFromStruct(Struct struct) {
```
Comment (gkalogiros-nutmeg) [inline src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: It feels that this belongs to a different class so that it can get tested properly

File: src/test/java/uk/co/threefi/connect/http/sink/HttpApiWriterTest.java [hunk 9]
```diff
@@ -574,6 +574,128 @@ public void testStructValue() throws Exception {
...
+    Schema valueSchema = SchemaBuilder.struct()
+            .field("id", Schema.STRING_SCHEMA)
+            .field("name", Schema.OPTIONAL_STRING_SCHEMA)
+            .build();
+
+    Struct structData = new Struct(valueSchema)
+            .put("id", "fake-user-id")
+            .put("name", "John Smith");
+
+    sinkRecords.add(new SinkRecord("user-topic",0,null,"fake-user-id",valueSchema, structData,0));
```
Comment (gkalogiros-nutmeg) [inline src/test/java/uk/co/threefi/connect/http/sink/HttpApiWriterTest.java]: Given that the main class has a few looks I'd like to see a test with mutliple records and keys... just in case

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java (hunk 1 above)
Comment (gastonNutmeg) [reply src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: fair point

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java (hunk 2 above)
Comment (gastonNutmeg) [reply src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: Actually, that gives a nullpointer because it needs to actually put it into the map. I tried with putIfAbsent and then using the result but the HashMap implementation of putIfAbsent is crazy....So I'm leaving it like that unless we can figure out another way

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java (hunk 4 above)
Comment (gastonNutmeg) [reply src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: moved

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java (hunk 5 above)
Comment (gastonNutmeg) [reply src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: didn't even notice the extra brackets. Refactored

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java (hunk 6 above)
Comment (gastonNutmeg) [reply src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: We need to publish all messages regardless so I think it'll be fine to send the empty keys, but yes ideally we should find another key that always have some sort of value (but the incoming key will do for now)

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java (hunk 7 above)
Comment (gastonNutmeg) [reply src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: Moved

File: src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java (hunk 8 above)
Comment (gastonNutmeg) [reply src/main/java/uk/co/threefi/connect/http/sink/HttpApiWriter.java]: For the time being I'd rather keep it here, but I will create a tech-debt ticket to move this out. Good call

File: src/test/java/uk/co/threefi/connect/http/sink/HttpApiWriterTest.java [hunk 10]
```diff
@@ -574,6 +574,128 @@ public void testStructValue() throws Exception {
...
+            .field("name", Schema.OPTIONAL_STRING_SCHEMA)
+            .field("test", Schema.OPTIONAL_STRING_SCHEMA)
+            .build();
+
+    Struct structData = new Struct(valueSchema)
+            .put("id", "fake-user-id")
+            .put("name", "John Smith")
+            .put("test", "To be removed");
+
+    sinkRecords.add(new SinkRecord("user-topic",0,null,"fake-user-id",valueSchema, structData,0));
```
[the remark this answers is not in the document]
Comment (gastonNutmeg) [reply src/test/java/uk/co/threefi/connect/http/sink/HttpApiWriterTest.java]: Done

File: src/test/java/uk/co/threefi/connect/http/sink/HttpApiWriterTest.java (hunk 9 above)
Comment (gastonNutmeg) [reply src/test/java/uk/co/threefi/connect/http/sink/HttpApiWriterTest.java]: Added a test for multiple keys and records
